core/node/libp2p/relay.go

```diff
@@ -70,3 +70,12 @@ func AutoRelay(addDefaultRelays bool) func() (opts Libp2pOpts, err error) {
 		return
 	}
 }
+
+func HolePunching(flag config.Flag, hasRelayClient bool) func() (opts Libp2pOpts, err error) {
+	return func() (opts Libp2pOpts, err error) {
+		if flag.WithDefault(false) hasRelayClient {
```

Comment:
@marten-seemann is this `flag.WithDefault(false) && hasRelayClient`? Should there be some error if the relay client is disabled, but holepunching is enabled?

Reply:
i think the libp2p constructor will complain then.

Reply:
It probably should, but currently it doesn't: https://github.com/libp2p/go-libp2p/blob/3a71a6d0591482de51b6eaa94093804ea21ab774/options.go#L457-L463

Makes me wonder if `EnableHolePunching` should be an option in the `RelayClient`: https://github.com/ipfs/go-ipfs-config/blob/71f0b2e315292164bcae5e2457fdfcc5225da759/swarm.go#L36-L45

Reply:
I think hole punching is a separate feature: there could be an implementation which does not depend on circle relay v2. Current config that keeps `EnableHolePunching` outside `RelayClient`  is fine. 

The question is: is there any utility _today_ in allowing `EnableHolePunching:true` while

Reply:
wait, relay client is autorelay; they are independent in libp2p, so it should be fine.